pkg/alertmanager/amcfg.go

```diff
@@ -305,57 +304,57 @@ func (cb *configBuilder) initializeFromRawConfiguration(b []byte) error {
-	for _, amConfigIdentifier := range util.SortedKeys(amConfigs) {
-		crKey := types.NamespacedName{
-			Name:      amConfigs[amConfigIdentifier].Name,
-			Namespace: amConfigs[amConfigIdentifier].Namespace,
-		}
-
-		// Add inhibitRules to baseConfig.InhibitRules.
-		for _, inhibitRule := range amConfigs[amConfigIdentifier].Spec.InhibitRules {
-			cb.cfg.InhibitRules = append(cb.cfg.InhibitRules,
-				cb.enforcer.processInhibitRule(
-					crKey,
-					cb.convertInhibitRule(
-						&inhibitRule,
-					),
-				),
-			)
-		}
+// addAlertmanagerConfig adds AlertmanagerConfig objects to the current configuration.
+func (cb *configBuilder) addAlertmanagerConfig(ctx context.Context, amConfig *monitoringv1alpha1.AlertmanagerConfig) error {
+	if amConfig == nil {
+		return nil
+	}
 
-		// Skip early if there's no route definition.
-		if amConfigs[amConfigIdentifier].Spec.Route == nil {
-			continue
-		}
+	subRoutes := make([]*route, 0, 1)
+	crKey := types.NamespacedName{
+		Name:      amConfig.Name,
+		Namespace: amConfig.Namespace,
+	}
 
-		subRoutes = append(subRoutes,
-			cb.enforcer.processRoute(
+	// Add inhibitRules to baseConfig.InhibitRules.
+	for _, inhibitRule := range amConfig.Spec.InhibitRules {
+		cb.cfg.InhibitRules = append(cb.cfg.InhibitRules,
+			cb.enforcer.processInhibitRule(
 				crKey,
-				cb.convertRoute(
-					amConfigs[amConfigIdentifier].Spec.Route,
-					crKey,
+				cb.convertInhibitRule(
+					&inhibitRule,
 				),
 			),
 		)
+	}
 
-		for _, receiver := range amConfigs[amConfigIdentifier].Spec.Receivers {
-			receivers, err := cb.convertReceiver(ctx, &receiver, crKey)
-			if err != nil {
-				return fmt.Errorf("AlertmanagerConfig %s: %w", crKey.String(), err)
-			}
-			cb.cfg.Receivers = append(cb.cfg.Receivers, receivers)
+	// Skip early if there's no route definition.
+	if amConfig.Spec.Route == nil {
+		return nil
+	}
+
+	subRoutes = append(subRoutes, cb.enforcer.processRoute(
+		crKey,
+		cb.convertRoute(
+			amConfig.Spec.Route,
+			crKey,
+		),
+	))
+
+	for _, receiver := range amConfig.Spec.Receivers {
+		receivers, err := cb.convertReceiver(ctx, &receiver, crKey)
+		if err != nil {
+			return fmt.Errorf("AlertmanagerConfig %s: %w", crKey.String(), err)
 		}
+		cb.cfg.Receivers = append(cb.cfg.Receivers, receivers)
+	}
 
-		for _, muteTimeInterval := range amConfigs[amConfigIdentifier].Spec.MuteTimeIntervals {
-			mti, err := convertMuteTimeInterval(&muteTimeInterval, crKey)
-			if err != nil {
-				return fmt.Errorf("AlertmanagerConfig %s: %w", crKey.String(), err)
-			}
-			cb.cfg.MuteTimeIntervals = append(cb.cfg.MuteTimeIntervals, mti)
+	for _, muteTimeInterval := range amConfig.Spec.MuteTimeIntervals {
+		mti, err := convertMuteTimeInterval(&muteTimeInterval, crKey)
+		if err != nil {
+			return fmt.Errorf("AlertmanagerConfig %s: %w", crKey.String(), err)
 		}
+		cb.cfg.MuteTimeIntervals = append(cb.cfg.MuteTimeIntervals, mti)
 	}
 
 	// For alerts to be processed by the AlertmanagerConfig routes, they need
```

Comment:
hmm this change is going to reverse the order of AlertmanagerConfig routes in the final configuration. It's an implementation detail but it might feel strange to see the routes listed in reverse order in the final configuration...
Also calling `cb.cfg.sanitize(cb.amVersion, cb.logger)` on every added AlertmanagerConfig isn't terribly efficient if we have lots of AlertmanagerConfigs but I'm not sure how/if we can avoid it.

Reply:
After discussing offline with @slashpai we realized that the change isn't going to do what we want because the invalid AlertmanagerConfig object needs to be added to the final configuration before we call cb.cfg.sanitize(). Thus while the latter will return an error (and the object be reported as rejected), it will still be part of the final configuration.
Looking into the different options we agreed that for now, we'll do a quick fix for the issue at hand and implement additional validation in sanitize() to ensure that SMTP host is defined either at the receiver level or at the global level. It will not be perfect because a single invalid AlertmanagerConfig will stop the reconciliation but it will avoid pushing invalid configs to Alertmanager.

Reply:
I have updated the old PR https://github.com/prometheus-operator/prometheus-operator/pull/6004 which fails the reconciliation due to invalid config